File: src/services/secrets.ts
/**
 * Secret storage.
 *
 * Blueprint §3.4: secrets live in Google Secret Manager / Cloud KMS; the app
 * DB holds only references. NOTE (assumption): this environment has no Google
 * Cloud, so the stand-in is a local vault: values are AES-256-GCM encrypted at
 * rest with a key derived from the AgentOS HMAC secret, stored in the
 * `secrets.value_enc` column. The `providerRef` column still records the
 * provider ("local-vault://<id>") so moving to Secret Manager is a swap of
 * this class only. Never store raw tokens in plaintext columns.
 */
import { randomBytes, createCipheriv, createDecipheriv, createHash } from "node:crypto";
import { randomUUID } from "node:crypto";
import { eq, and } from "drizzle-orm";
import { secrets } from "../db/schema.js";
import type { DB } from "../db/client.js";
import type { SecretRef, SecretPurpose } from "../domain/types.js";

export class SecretService {
  constructor(
    private db: DB,
    private hmacSecret: string,
  ) {}

  private key(): Buffer {
    return createHash("sha256").update("agentos-vault:" + this.hmacSecret).digest();
  }

  encrypt(plain: string): string {
    const iv = randomBytes(12);
    const cipher = createCipheriv("aes-256-gcm", this.key(), iv);
    const enc = Buffer.concat([cipher.update(plain, "utf8"), cipher.final()]);
    const tag = cipher.getAuthTag();
    return [iv.toString("base64url"), tag.toString("base64url"), enc.toString("base64url")].join(".");
  }

  decrypt(payload: string): string {
    const [ivB, tagB, encB] = payload.split(".");
    const decipher = createDecipheriv("aes-256-gcm", this.key(), Buffer.from(ivB!, "base64url"));
    decipher.setAuthTag(Buffer.from(tagB!, "base64url"));
    return Buffer.concat([decipher.update(Buffer.from(encB!, "base64url")), decipher.final()]).toString("utf8");
  }

  async createRef(
    projectId: string,
    name: string,
    purpose: SecretPurpose,
    value?: string,
  ): Promise<SecretRef> {
    const id = randomUUID();
    const row: SecretRef = {
      id,
      projectId,
      name,
      providerRef: `local-vault://${id}`,
      purpose,
    };
    await this.db
      .insert(secrets)
      .values({
        ...row,
        valueEnc: value !== undefined ? this.encrypt(value) : null,
        createdAt: new Date().toISOString(),
      })
      .run();
    return row;
  }

  async setValue(id: string, value: string): Promise<void> {
    await this.db
      .update(secrets)
      .set({ valueEnc: this.encrypt(value) })
      .where(eq(secrets.id, id))
      .run();
  }

  /** Returns the decrypted value or null if the ref has no value stored. */
  async getValue(id: string): Promise<string | null> {
    const row = await this.db.select().from(secrets).where(eq(secrets.id, id)).get();
    if (!row || !row.valueEnc) return null;
    try {
      return this.decrypt(row.valueEnc);
    } catch {
      return null;
    }
  }

  async list(projectId: string): Promise<SecretRef[]> {
    const rows = await this.db.select().from(secrets).where(eq(secrets.projectId, projectId)).all();
    return rows.map((r) => ({
      id: r.id,
      projectId: r.projectId,
      name: r.name,
      providerRef: r.providerRef,
      purpose: r.purpose,
    }));
  }

  async delete(id: string): Promise<void> {
    await this.db.delete(secrets).where(eq(secrets.id, id)).run();
  }
}

File: agentos/services/secrets.py
"""
Secret storage service. Port of src/services/secrets.ts.
Values are AES-256-GCM encrypted at rest; DB holds only references.
"""
import base64
import hashlib
import os
import uuid
from datetime import datetime, timezone

from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from sqlalchemy import select, delete
from sqlalchemy.ext.asyncio import AsyncSession

from agentos.db.client import get_session
from agentos.db.models import Secret as SecretRow


class SecretService:
    def __init__(self, hmac_secret: str) -> None:
        self._key = hashlib.sha256(f"agentos-vault:{hmac_secret}".encode()).digest()

    def encrypt(self, plain: str) -> str:
        nonce = os.urandom(12)
        aesgcm = AESGCM(self._key)
        ct = aesgcm.encrypt(nonce, plain.encode(), None)
        # ct includes the 16-byte GCM tag appended by cryptography
        return (
            base64.urlsafe_b64encode(nonce).rstrip(b"=").decode()
            + "."
            + base64.urlsafe_b64encode(ct).rstrip(b"=").decode()
        )

    def decrypt(self, payload: str) -> str:
        nonce_b, ct_b = payload.split(".", 1)
        nonce = base64.urlsafe_b64decode(nonce_b + "==")
        ct = base64.urlsafe_b64decode(ct_b + "==")
        aesgcm = AESGCM(self._key)
        return aesgcm.decrypt(nonce, ct, None).decode()

    async def create_ref(self, project_id: str, name: str, purpose: str,
                         value: str | None = None) -> dict:
        ref_id = str(uuid.uuid4())
        row = SecretRow(
            id=ref_id,
            project_id=project_id,
            name=name,
            provider_ref=f"local-vault://{ref_id}",
            purpose=purpose,
            value_enc=self.encrypt(value) if value is not None else None,
            created_at=datetime.now(timezone.utc).isoformat(),
        )
        async with get_session() as db:
            db.add(row)
            await db.commit()
        return {"id": ref_id, "projectId": project_id, "name": name,
                "providerRef": row.provider_ref, "purpose": purpose}

    async def set_value(self, ref_id: str, value: str) -> None:
        async with get_session() as db:
            result = await db.execute(select(SecretRow).where(SecretRow.id == ref_id))
            row = result.scalar_one_or_none()
            if row:
                row.value_enc = self.encrypt(value)
                await db.commit()

    async def get_value(self, ref_id: str) -> str | None:
        async with get_session() as db:
            result = await db.execute(select(SecretRow).where(SecretRow.id == ref_id))
            row = result.scalar_one_or_none()
        if not row or not row.value_enc:
            return None
        try:
            return self.decrypt(row.value_enc)
        except Exception:
            return None

    async def list(self, project_id: str) -> list[dict]:
        async with get_session() as db:
            result = await db.execute(select(SecretRow).where(SecretRow.project_id == project_id))
            rows = result.scalars().all()
        return [{"id": r.id, "projectId": r.project_id, "name": r.name,
                 "providerRef": r.provider_ref, "purpose": r.purpose} for r in rows]

    async def delete(self, ref_id: str) -> None:
        async with get_session() as db:
            await db.execute(delete(SecretRow).where(SecretRow.id == ref_id))
            await db.commit()
